### ISSUE TITLE
Utilize array instead of vector in C

### ISSUE THREAD

**Issue by krahets** [krahets/hello-algo#869]
Hi, @Reanon @Gonglja @lhxsm  !

Currently we utilize self-implemented `vector` in `utils/vector.h` trying to align with C++ code. However, I recommend depreciating `vector` for these reasons:

1. We try to make readers be available to solve algorithm problems using C. Usually, they will not have a chance to access the `vector` when solving the problems because it is not a built-in struct.
2. The code using `vector` seems like more complex than built-in array. For instance:
    1. `bool select = *((bool *)(selected->data[I]));\n if (!select)` --> `if (!selected[i])`
    2. *((bool *)selected->data[i]) = true` --> `selected[i] = true`

I've re-implemented `permutations_i.c`in #868 , Could you please update other source codes? Thanks!

<img width="297" alt="image" src="https://github.com/krahets/hello-algo/assets/26993056/f3be1db5-7a58-4a13-8316-29190f775aeb"> 

Before:

```c
/* 回溯算法：全排列 I */
void backtrack(vector *state, vector *choices, vector *selected, vector *res) {
    // 当状态长度等于元素数量时，记录解
    if (state->size == choices->size) {
        vector *newState = newVector();
        for (int i = 0; i < state->size; i++) {
            vectorPushback(newState, state->data[i], sizeof(int));
        }
        vectorPushback(res, newState, sizeof(vector));
        return;
    }
    // 遍历所有选择
    for (int i = 0; i < choices->size; i++) {
        int *choice = malloc(sizeof(int));
        *choice = *((int *)(choices->data[i]));
        // 剪枝：不允许重复选择元素 且 不允许重复选择相等元素
        bool select = *((bool *)(selected->data[i]));
        if (!select) {
            // 尝试：做出选择，更新状态
            *((bool *)selected->data[i]) = true;
            vectorPushback(state, choice, sizeof(int));
            // 进行下一轮选择
            backtrack(state, choices, selected, res);
            // 回退：撤销选择，恢复到之前的状态
            *((bool *)selected->data[i]) = false;
            vectorPopback(state);
        }
    }
}

/* 全排列 I */
vector *permutationsI(vector *nums) {
    vector *iState = newVector();

    int select[3] = {false, false, false};
    vector *bSelected = newVector();
    for (int i = 0; i < nums->size; i++) {
        vectorPushback(bSelected, &select[i], sizeof(int));
    }

    vector *res = newVector();

    // 前序遍历
    backtrack(iState, nums, bSelected, res);
    return res;
}

/* 打印向量中的元素 */
void printFunc(vector *v, void *p) {
    int *node = p;
    printf("%d", *node);
}

/* Driver Code */
int main() {
    int nums[] = {1, 2, 3};
    vector *iNums = newVector(); // int
    for (int i = 0; i < sizeof(nums) / sizeof(nums[0]); i++) {
        vectorPushback(iNums, &nums[i], sizeof(int));
    }

    vector *res = permutationsI(iNums);

    // 输出结果
    printf("输入数组 nums = ");
    printArray(nums, sizeof(nums) / sizeof(nums[0]));
    printf("所有排列 res = ");
    printVectorMatrix(res, printFunc);

    // 释放内存
    delVector(iNums);
    delVector(res);
    return 0;
}
```

After:

```c
// 假设最多有 1000 个排列
#define MAX_SIZE 1000

/* 回溯算法：全排列 I */
void backtrack(int *state, int stateSize, int *choices, int choicesSize, bool *selected, int **res, int *resSize) {
    // 当状态长度等于元素数量时，记录解
    if (stateSize == choicesSize) {
        res[*resSize] = (int *)malloc(choicesSize * sizeof(int));
        for (int i = 0; i < choicesSize; i++) {
            res[*resSize][i] = state[i];
        }
        (*resSize)++;
        return;
    }
    // 遍历所有选择
    for (int i = 0; i < choicesSize; i++) {
        int choice = choices[i];
        // 剪枝：不允许重复选择元素
        if (!selected[i]) {
            // 尝试：做出选择，更新状态
            selected[i] = true;
            state[stateSize] = choice;
            // 进行下一轮选择
            backtrack(state, stateSize + 1, choices, choicesSize, selected, res, resSize);
            // 回退：撤销选择，恢复到之前的状态
            selected[i] = false;
        }
    }
}

/* 全排列 I */
int **permutationsI(int *nums, int numsSize, int *returnSize) {
    int *state = (int *)malloc(numsSize * sizeof(int));
    bool *selected = (bool *)malloc(numsSize * sizeof(bool));
    for (int i = 0; i < numsSize; i++) {
        selected[i] = false;
    }
    int **res = (int **)malloc(MAX_SIZE * sizeof(int *));
    *returnSize = 0;

    backtrack(state, 0, nums, numsSize, selected, res, returnSize);

    free(state);
    free(selected);

    return res;
}

/* Driver Code */
int main() {
    int nums[] = {1, 2, 3};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int returnSize;

    int **res = permutationsI(nums, numsSize, &returnSize);

    printf("输入数组 nums = ");
    printArray(nums, numsSize);
    printf("\n所有排列 res = \n");
    for (int i = 0; i < returnSize; i++) {
        printArray(res[i], numsSize);
    }

    // 释放内存
    for (int i = 0; i < returnSize; i++) {
        free(res[i]);
    }
    free(res);

    return 0;
}
```

**Comment by lhxsm** [comment]
Hi @krahets ,Thx for the recognition.
I'm not quite sure if I have the extra time, and I having traveled outside recently. But I'll help out with it when I have time.

**Comment by Gonglja** [comment]
Hi, I've been on a business trip recently. I'm expected to return next Tuesday. I'll update it when I have time.

发自我的iPhone

在 2023年10月17日，23:38，Yudong Jin ***@***.***> 写道：

﻿

Hi, @Gonglja<https://github.com/Gonglja> @lhxsm<https://github.com/lhxsm>!

Currently we utilize self-implemented vector in utils/vector.h trying to align with C++ code. However, I recommend depreciating vector for these reasons:

  1.  We try to make readers be available to solve algorithm problems using C. Usually, they will not have a chance to access the vector when solving the problems because it is not a built-in struct.
  2.  The code using vector seems like more complex than built-in array. For instance:
     *   bool select = *((bool *)(selected->data[I]));\n if (!select) --> if (!selected[i])
     *   *((bool *)selected->data[i]) = true-->selected[i] = true`

I've re-implemented permutations_i.cin #868<https://github.com/krahets/hello-algo/pull/868> , Could you please update other source codes? Thanks!

[image]<https://user-images.githubusercontent.com/26993056/275893260-f3be1db5-7a58-4a13-8316-29190f775aeb.png>

Before:

/* 回溯算法：全排列 I */
void backtrack(vector *state, vector *choices, vector *selected, vector *res) {
    // 当状态长度等于元素数量时，记录解
    if (state->size == choices->size) {
        vector *newState = newVector();
        for (int i = 0; i < state->size; i++) {
            vectorPushback(newState, state->data[i], sizeof(int));
        }
        vectorPushback(res, newState, sizeof(vector));
        return;
    }
    // 遍历所有选择
    for (int i = 0; i < choices->size; i++) {
        int *choice = malloc(sizeof(int));
        *choice = *((int *)(choices->data[i]));
        // 剪枝：不允许重复选择元素 且 不允许重复选择相等元素
        bool select = *((bool *)(selected->data[i]));
        if (!select) {
            // 尝试：做出选择，更新状态
            *((bool *)selected->data[i]) = true;
            vectorPushback(state, choice, sizeof(int));
            // 进行下一轮选择
            backtrack(state, choices, selected, res);
            // 回退：撤销选择，恢复到之前的状态
            *((bool *)selected->data[i]) = false;
            vectorPopback(state);
        }
    }
}

/* 全排列 I */
vector *permutationsI(vector *nums) {
    vector *iState = newVector();

    int select[3] = {false, false, false};
    vector *bSelected = newVector();
    for (int i = 0; i < nums->size; i++) {
        vectorPushback(bSelected, &select[i], sizeof(int));
    }

    vector *res = newVector();

    // 前序遍历
    backtrack(iState, nums, bSelected, res);
    return res;
}

/* 打印向量中的元素 */
void printFunc(vector *v, void *p) {
    int *node = p;
    printf("%d", *node);
}

/* Driver Code */
int main() {
    int nums[] = {1, 2, 3};
    vector *iNums = newVector(); // int
    for (int i = 0; i < sizeof(nums) / sizeof(nums[0]); i++) {
        vectorPushback(iNums, &nums[i], sizeof(int));
    }

    vector *res = permutationsI(iNums);

    // 输出结果
    printf("输入数组 nums = ");
    printArray(nums, sizeof(nums) / sizeof(nums[0]));
    printf("所有排列 res = ");
    printVectorMatrix(res, printFunc);

    // 释放内存
    delVector(iNums);
    delVector(res);
    return 0;
}

After:

// 假设最多有 1000 个排列
#define MAX_SIZE 1000

/* 回溯算法：全排列 I */
void backtrack(int *state, int stateSize, int *choices, int choicesSize, bool *selected, int **res, int *resSize) {
    // 当状态长度等于元素数量时，记录解
    if (stateSize == choicesSize) {
        res[*resSize] = (int *)malloc(choicesSize * sizeof(int));
        for (int i = 0; i < choicesSize; i++) {
            res[*resSize][i] = state[i];
        }
        (*resSize)++;
        return;
    }
    // 遍历所有选择
    for (int i = 0; i < choicesSize; i++) {
        int choice = choices[i];
        // 剪枝：不允许重复选择元素
        if (!selected[i]) {
            // 尝试：做出选择，更新状态
            selected[i] = true;
            state[stateSize] = choice;
            // 进行下一轮选择
            backtrack(state, stateSize + 1, choices, choicesSize, selected, res, resSize);
            // 回退：撤销选择，恢复到之前的状态
            selected[i] = false;
        }
    }
}

/* 全排列 I */
int **permutationsI(int *nums, int numsSize, int *returnSize) {
    int *state = (int *)malloc(numsSize * sizeof(int));
    bool *selected = (bool *)malloc(numsSize * sizeof(bool));
    for (int i = 0; i < numsSize; i++) {
        selected[i] = false;
    }
    int **res = (int **)malloc(MAX_SIZE * sizeof(int *));
    *returnSize = 0;

    backtrack(state, 0, nums, numsSize, selected, res, returnSize);

    free(state);
    free(selected);

    return res;
}

/* Driver Code */
int main() {
    int nums[] = {1, 2, 3};
    int numsSize = sizeof(nums) / sizeof(nums[0]);
    int returnSize;

    int **res = permutationsI(nums, numsSize, &returnSize);

    printf("输入数组 nums = ");
    printArray(nums, numsSize);
    printf("\n所有排列 res = \n");
    for (int i = 0; i < returnSize; i++) {
        printArray(res[i], numsSize);
    }

    // 释放内存
    for (int i = 0; i < returnSize; i++) {
        free(res[i]);
    }
    free(res);

    return 0;
}

**Comment by krahets** [comment]
Fixed in #894
